### ISSUE TITLE
Does sly-cl-indent have to provide cl-indent?

### ISSUE THREAD

**Issue by tarsius** [joaotavora/sly#92]
Maybe it does, e.g. if `sly-cl-indent` is considered to be a drop in replacement for `cl-indent`, then doing so _might_ be a reasonable course of action. (But that would only be the case if using the sly variant doesn't mess with other packages that depend on `cl-indent`.)

If you don't want to stop providing that feature, that's fine. But I need to know so that I can write down the reason for the manual patch to the database needed to prevent the issue this causes for users of `epkg.el`+`borg.el`. (A package which requires `cl-indent` will then depend on `slime`, currently the dependency is undefined, either `slime` or `sly`.

**Comment by tarsius** [comment]
By the way `cl-indent` is now part of Emacs (as of 25.1) and is also available from GNU Elpa.

**Comment by joaotavora** [comment]
I have to take a better look at this, I would really love to ditch `sly-cl-indent` altogether...

**Comment by tarsius** [comment]
Friendly ping!

**Comment by tarsius** [comment]
> I would really love to ditch sly-cl-indent altogether

That sounds like a good plan.

**Comment by joaotavora** [comment]
> That sounds like a good plan.

I've had a better look and that is very hard. Only `sly-cl-indent` correctly manages that special `loop` indentation that I and many lispers can't live without. It appears `cl-indent.el` used to do this, but it doesn't anymore.  I need to investigate this further. There is a way out, but it's basically throwing in the towel and probably also hard: Rename every symbol in `sly-cl-indent.el` that conflicts with `cl-indent.el`, and stop providing `cl-indent`

**Comment by tarsius** [comment]
Thanks for looking into this. I hope you can find a suitable and not too laborious solution, but I will try not to bug you about it anymore ;-)

**Comment by joaotavora** [comment]
Yeah, i fear it's quite laborious.... Another option would be to get sly-cl-indent.el merged into emacs and have it replace `cl-indent`

These are its current authors according to a `git log`. Perhaps many of them have already signed copyright assignments or have otherwise contributed trivial patches. Didier Verna is apparently also there. 
```
      2:Author: Deokhwan Kim <deokhwankim@gmail.com>
     11:Author: Luís Oliveira <loliveira@common-lisp.net>
     20:Author: Joao Tavora <joaotavora@gmail.com>
    114:Author: Samuel Freilich <sfreilich@google.com>
    120:Author: Stas Boukarev <sboukarev@common-lisp.net>
    127:Author: Nikodemus Siivola <nsiivola@common-lisp.net>
```

**Comment by joaotavora** [comment]
(By the way, the commit that references this has nothing to do with this, it's a headless commit meant to reference 91...)

**Comment by tarsius** [comment]
I am closing my old issues.

**Comment by joaotavora** [comment]
> I am closing my old issues.

And I am reopening it, since it is still a real problem in SLY :-)

**Comment by cireu** [comment]
Another problem is, `cl-indent` allow you use different indentation for Elisp, by setting property `common-lisp-indent-function-for-elisp` of symbol `common-lisp-indent-function`. `sly-cl-indent` don't support this. It means if someone use `common-lisp-indent-function` to indent Elisp code(Following `cl-indent`'s usage), their Elisp indentation may get broken after load `sly`.

**Comment by tarsius** [comment]
> There is a way out, but it's basically throwing in the towel and probably also hard: Rename every symbol in sly-cl-indent.el that conflicts with cl-indent.el, and stop providing cl-indent

At this point that's what I think you should do. (Except that you should rename *each and every* symbol.)

**Comment by SwiftLawnGnome** [comment]
I've done this in my own [fork](https://github.com/SwiftLawnGnome/sly), it wasn't as tedious as it seemed. Been using it for almost a month without issues. I didn't totally streamline the naming conventions (though all symbols do now start with `sly`) but that's easy to change. I'll happily submit a PR if you'd like, this has been my biggest pet peeve with Sly/Slime as long as I've used them.

**Comment by joaotavora** [comment]
> I'll happily submit a PR if you'd like, this has been my biggest pet peeve with Sly/Slime as long as I've used them.

Sure @SwiftLawnGnome , I think that's a good idea, at least to have a starting point.  Though a naming convention is useful.

**Comment by joaotavora** [comment]
@SwiftLawnGnome , I've had a look at your fork.  Pretty large diff.  You say it wasn't tedious?!! That's great! The naming convention you adopted doesn't seem terrible at all.  Maybe, if you can manage, I would replace all the `sly-` with `sly--` so we don't conceptually export anything outside of SLY, for now at least.  But other than that, it looks pretty good.  I would definitely like to have a pull request for this, so I can test it for a while.

Thanks very much

**Comment by SwiftLawnGnome** [comment]
> I've had a look at your fork. Pretty large diff. You say it wasn't tedious?!!

Well, it _was_ tedious, but iedit and counsel-rg kept me sane!

> Maybe, if you can manage, I would replace all the `sly-` with `sly--` so we don't conceptually export anything outside of SLY

I just pushed a bunch of commits which changed most symbols. The `sly--lisp-indent-` prefix is used almost everywhere. Some external symbols may have the `-` that shouldn't and vice versa, let me know of any other naming changes you think are appropriate. I haven't tried these recent changes yet, and considering how many there are, there's a very good chance something's broken. I probably won't be able to do much testing atm but I'll go ahead and submit the PR.

**Comment by joaotavora** [comment]
I finally did the reviewing work and pushed this.  @tarsius we have @SwiftLawnGnome to thank for this.

@SwiftLawnGnome , thanks very much: you did a lot of hard work, renaming and making everything consistent in the `sly-indent-cl.el` library and its users. It took me some time to review the most important commits, but I selected those and pushed them to master.  I went through basically every little decision you took and agreed with _all_ of them, no exception.  You did introduce a minor bug with the "indent-13" test case that took me a while to figure out, but in the end I found it fixed it.

In #312 you have a bunch of unrelated, but also very interesting commits that I will push to master incrementally, as I review them.  I will close that Pull Request, but don't remove that branch, as I will be cherry-picking the commits.
